### PR TITLE
Import from URL action 

### PR DESCRIPTION
**Pull request by nirvn**
*A picture*
image

*An explanation*
This PR add a new 'import URL' action in the local data picker screen, which allows for users to paste a URL (say a dropbox link) and have QField automatically download and import the dataset or zipped project.

This is useful across Android and iOS, but comes in especially handy for iOS as it has currently no other mean to import datasets and projects nor file association. Without this, users have to do a fair amount of acrobatics when trying to import within the phone directly.

*Why ZipUtils?*
It's there until the ZipUtils::files() and the new parameter in ZipUtils::unzip() is upstreamed.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for c971fc27499d81cf55f3d8ab816227638a4973c5: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-arm64-android.apk)
- [all access arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-arm-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4027-2423ed5e5814e1909845fd917f0f72a889d8f514-x86-android.apk)
</details>

File: src/qml/QFieldLocalDataPickerScreen.qml [hunk 1]
```diff
@@ -519,6 +538,46 @@ Page {
...
+      QfTextField {
+        id: importUrlInput
+        width: importUrlLabel.width
+        focus: true
+      }
+    }
+
+    standardButtons: Dialog.Ok | Dialog.Cancel
+    onAccepted: {
+      form.cancel()
```
Comment (nirvn) [inline src/qml/QFieldLocalDataPickerScreen.qml:577]:
```suggestion
      iface.importUrl(importUrlLabel.text)
```

File: src/qml/QFieldLocalDataPickerScreen.qml [hunk 2]
```diff
@@ -519,6 +538,46 @@ Page {
...
+      QfTextField {
+        id: importUrlInput
+        width: importUrlLabel.width
+        focus: true
+      }
+    }
+
+    standardButtons: Dialog.Ok | Dialog.Cancel
+    onAccepted: {
+      iface.importUrl(importUrlLabel.text)
```
Comment (nirvn) [inline src/qml/QFieldLocalDataPickerScreen.qml:577]:
```suggestion
      iface.importUrl(importUrlInput.text)
```

File: src/core/appinterface.h [hunk 3]
```diff
@@ -38,6 +38,8 @@ class AppInterface : public QObject
       Q_ASSERT( false );
     }
 
+    Q_INVOKABLE void importUrl( const QString &url );
```
Comment (m-kuhn) [inline src/core/appinterface.h:41]: Please add some docs :-)

File: src/core/appinterface.h [hunk 4]
```diff
@@ -90,6 +92,10 @@ class AppInterface : public QObject
   signals:
     void openFeatureFormRequested();
 
+    void importUrlTriggered( const QString &url, const QString &name );
+
+    void importUrlEnded( const QString &path = QString() );
```
Comment (m-kuhn) [inline src/core/appinterface.h:97]: Also here

**Review by m-kuhn — APPROVED** [review]

File: src/core/appinterface.h (hunk 3 above)
Comment (nirvn) [reply src/core/appinterface.h:41]: Done, added coverage for other doc-less signals too.

**Comment by qfield-fairy** [comment]
The backport to `release-2_6` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply 94c233b54... Generalize busyMessage overlay, move into its own QML file
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging src/qml/Changelog.qml
Auto-merging src/qml/qgismobileapp.qml
CONFLICT (content): Merge conflict in src/qml/qgismobileapp.qml
Auto-merging src/qml/qml.qrc

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-release-2_6 release-2_6
# Navigate to the new working tree
cd .worktrees/backport-release-2_6
# Create a new branch
git switch --create backport-4027-to-release-2_6
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 94c233b541e15cee2f6df4fa54a2e546924da7e7,a687931823c9af7e7c639176a5d11b4ef8e3ce70,9dba89cf6df16ed5fa6f7dd8f4a994d838d6c81b,eb0172ca838171563f856193525aa9692c9287b9,6d8216db858230cbfc60e975c69ecd7a5b98d5e5,a77a737eb6c23707b9be7e7ec74d81e4c2e82f0a,9ef5c3c2db08fafe2719ddf3eb31242515b799ac,954d304ba5b2949b9433cbc9c9ec6041820d55ca
# Push it to GitHub
git push --set-upstream origin backport-4027-to-release-2_6
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-release-2_6
```
Then, create a pull request where the `base` branch is `release-2_6` and the `compare`/`head` branch is `backport-4027-to-release-2_6`.

**Comment by nirvn** [comment]
Oups, we're 2.7 already.

**Comment by TomKenda** [comment]
[Not abble to import from URL]

***Describe the issue***
When using the new Import from URL function, when downloading a .zip file from:
 - Dropbox : a project is open in Qfield but this project is blank, no layer are displayed
 -  OneDrive : nothing happen when I copy paste the url and click on "ok"

***Reproduction steps***
Steps to reproduce the behavior:

- Open Import from URL dialog box
- Paste in URL from 'Sample data' below to test
- click "ok"

***Expected behavior***
.zip file with .qgs file should be able to be opened as a project.

***Observed behavior***
- OneDrive : Fille is not opened, nothing happens, I am just redirected to the "imported datasets" folder of the app where my folder does not appear.
- Dropbox : the name of the zipped folder is displayed in the "imported datasets" folder of the app. I can then click on the file to open the project in Qfield. But When I do so, everything is blank, no layer are displayed in the project and nothing appears on the map.


***Sample data***
Project with subfolders DropBox: [QField collect 1](https://www.dropbox.com/s/sx2wprp7k43mvf1/QField_projet_direct_folders.zip?dl=0)

Project with no subfolders (flat folder structure) Dropbox: [QField collect 2](https://www.dropbox.com/s/7rc68b25d77ixyp/QField_projet_direct_files.zip?dl=0)


Project with subfolders OneDrive: [QField collect 3](https://cragx-my.sharepoint.com/:u:/g/personal/t_kenda_cra_wallonie_be/EcpdVJGa2nJGtd5MzvmzDBABLmM3Yi8urAVilMpVCPrkKQ?e=yaroWq)


***Mobile***
Device: iPhone 8
OS: iOS 16.4.1
QField version: 3.30


Additional information
\

**Comment by nirvn** [comment]
@TomKenda , you have to make sure the URL points to the actual file, not an html landing page 

For eg, with Dropbox, make sure the URL dl parameter is set to 1:

`https://www.dropbox.com/s/sx2wprp7k43mvf1/QField_projet_direct_folders.zip?dl=1`

**Pull request merged**